libs/auth/src/dipped_smart_card_auth.ts

```diff
@@ -384,97 +384,93 @@ export class DippedSmartCardAuth implements DippedSmartCardAuthApi {
                 }
+                return {
+                  status: 'logged_out',
+                  reason: validationResult.err(),
+                  cardUserRole: user?.role,
+                };
               }
-            }
 
-            case 'checking_pin': {
-              if (action.cardStatus.status === 'no_card') {
-                return { status: 'logged_out', reason: 'machine_locked' };
+              /* istanbul ignore next: Compile-time check for completeness */
+              default: {
+                return throwIllegalValue(action.cardStatus, 'status');
               }
-              return currentAuthStatus;
             }
+          }
 
-            case 'remove_card': {
-              if (action.cardStatus.status === 'no_card') {
-                const { user } = currentAuthStatus;
-                if (user.role === 'system_administrator') {
-                  return {
-                    status: 'logged_in',
-                    user,
-                    programmableCard: cardStatusToProgrammableCard(
-                      action.cardStatus
-                    ),
-                  };
-                }
-                return { status: 'logged_in', user };
-              }
-              return currentAuthStatus;
+          case 'checking_pin': {
+            if (action.cardStatus.status === 'no_card') {
+              return { status: 'logged_out', reason: 'machine_locked' };
             }
+            return currentAuthStatus;
+          }
 
-            case 'logged_in': {
+          case 'remove_card': {
+            if (action.cardStatus.status === 'no_card') {
               const { user } = currentAuthStatus;
               if (user.role === 'system_administrator') {
                 return {
-                  ...currentAuthStatus,
+                  status: 'logged_in',
+                  user,
                   programmableCard: cardStatusToProgrammableCard(
                     action.cardStatus
                   ),
                 };
               }
-              return currentAuthStatus;
+              return { status: 'logged_in', user };
             }
+            return currentAuthStatus;
+          }
 
-            /* istanbul ignore next: Compile-time check for completeness */
-            default: {
-              throwIllegalValue(currentAuthStatus, 'status');
+          case 'logged_in': {
+            const { user } = currentAuthStatus;
+            if (user.role === 'system_administrator') {
+              return {
+                ...currentAuthStatus,
+                programmableCard: cardStatusToProgrammableCard(
+                  action.cardStatus
+                ),
+              };
             }
+            return currentAuthStatus;
           }
-        })();
 
-        return newAuthStatus;
+          /* istanbul ignore next: Compile-time check for completeness */
+          default: {
+            return throwIllegalValue(currentAuthStatus, 'status');
```

Comment:
Slight preference for leaving this without the `return` and putting a `break` in the outer `switch` `case`.

Reply:
I also prefer a `break`, but using a `break` here results in TS detecting unreachable code:

<img width="710" alt="unreachable-code" src="https://user-images.githubusercontent.com/12616928/225190931-3d2b8120-c116-40f2-b42b-2fc686532a31.png">

And for full visibility, just removing the `return` isn't an option either since that interferes with eslint's no-fallthrough rule:

<img width="700" alt="eslint" src="https://user-images.githubusercontent.com/12616928/225191848-1cab1b5b-cb68-4d95-818c-8c02c63a2203.png">

So a bit of a catch-22 haha